### DISCUSSION TITLE
Question on custom messages

### DISCUSSION THREAD

**Discussion by 7omg** [Textualize/textual#1990] [Q&A] [answered]
Hi, 
I'm doing my first project with textual and I have a very hard time understanding how the messaging system works.
The application  has to lists. When an entry in the first list is selected there should be a message sent to the other list updating it with the data from the first list entry. (Sorry if this is a stupid approach :')
Simply said: My goal is to message a widget.
Below you can see a minimal example of what I'm trying to achieve.

```python
from textual import log
from textual.app import App, ComposeResult
from textual.message import Message, MessageTarget
from textual.widgets import Static, ListView, ListItem


class SomeList(ListView):
    """First list"""

    class Update(Message):
        """Custom message used to send data to another widget"""
        def __init__(self, sender: MessageTarget, data: str):
            super().__init__(sender)
            self.data = data

    def compose(self) -> ComposeResult:
        yield ListItem(Static("Entry one"))
        yield ListItem(Static("Entry two"))
        yield ListItem(Static("Entry three"))

    def on_list_view_selected(self, message):
        log(f"SomeList.on_list_view_selected")

        # post an `Update` message with the string of the static inside as data attribute
        self.post_message(self.Update(self, message.item.children[0].renderable))


class AnotherList(ListView):
    """Second list"""

    def compose(self) -> ComposeResult:
        yield ListItem(Static("Hello"))

    def on_some_list_update(self, message: SomeList.Update):
        log("AnotherList.on_some_list_update")
        self.append(ListItem(Static(f"New: {message.data}")))


class MyApp(App):
    def compose(self) -> ComposeResult:
        yield Static("FIRST LIST")
        yield SomeList()

        yield Static("SECOND LIST")
        yield AnotherList()

    def on_some_list_update(self):
        """This should be used to forward the message to `AnotherList`"""
        log("MyApp.on_some_list_update")

    def on_list_view_update(self):
        """Same as above as I don't know which name is the correct one"""
        log("MyApp.on_list_view_update")


if __name__ == "__main__":
    app = MyApp()
    app.run()

```
**First Question:**
I read that it's common to 'message yourself' as done in `SomeList.on_list_view_selected()`. This should have a sort of 'bubbling up' effect where the message is passed to the parent widgets. My intention was to catch this bubbled up message in `MyApp.on_some_list_update()` (or `MyApp.on_list_view_update()`?) and pass it on to `AnotherList.on_list_view_selected()`. However the log message in both of these methods is never printed when I run with `textual console` and `textual run --dev main.py`. Why is that?

**Second Question:**
Every time I start the app and select a entry from the first list I get the following log:
```text
[01:52:42] INFO                                                                                              main.py:23
SomeList.on_list_view_selected
[01:52:42] PRINT                                                                                            warnings.py:30
/path/to/main.py:26: RuntimeWarning: coroutine 'Widget.post_message' was never awaited
  self.post_message(self.Update(self, message.item.children[0].renderable))
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```
I expected the first log message but don't know what it means that `Widget.post_message` was _never awaited_. I know it has something to do with asynchronous behavior but can't find a helpful resource.

Thanks in advance :)

**Comment by willmcgugan** [comment]
There was a change to post_message recently. The code looks fine, but the error suggests you have an old version of Textual. Try updating, you might find your example works as expected.

**Comment by davep** [comment, marked as answer]
Further to what Will says above, I think you may also have run into a small issue we do have from time to time: it's not uncommon that the documentation for the `main` branch "leaks" onto the website before the next release is made. From what I can see of your problem you're (rightly, as per the docs right now) calling `post_message` when you likely really want `post_message_no_wait`; this is assuming you're working with 0.13 of Textual rather than a clone of `main`.

For the moment things should work for you if you use `post_message_no_wait`; but note that [this will change in 0.14](https://github.com/Textualize/textual/blob/main/CHANGELOG.md#0140---unreleased), where it *will* be `post_message` and you won't need to `await` it.

Hope that makes sense?

Long story short: one of those rare times where reading the fine manual caused the problem, and it's our fault. ;-)

**Reply by 7omg** [reply]
Thank you so much Dave. This problem was driving me crazy. Is it already known when 0.14 will be released?

**Reply by davep** [reply]
Maaaaaaaaaybe today; but don't quote me on it. ;-)

**Reply by davep** [reply]
Just to update, 0.14 has escaped: https://textual.textualize.io/blog/2023/03/09/textual-0140-shakes-up-posting-messages/
